litert/vendors/qualcomm/core/wrappers/tests/tensor_wrapper_test.cc

```diff
@@ -457,61 +458,124 @@ TEST(TensorWrapperTest, ConstQnnTensorPerTensorQuantConstructTest) {
+                                    tensor.v2.dimensions + tensor.v2.rank),
+              ::testing::ElementsAreArray(expected_dims));
+
+  // Validate data buffer.
+  ASSERT_EQ(tensor.v2.clientBuf.dataSize, expected_data.size() * sizeof(T));
+  const auto* tensor_data_ptr = static_cast<const T*>(tensor.v2.clientBuf.data);
+  EXPECT_THAT(
+      std::vector<T>(tensor_data_ptr, tensor_data_ptr + expected_data.size()),
+      ::testing::ElementsAreArray(expected_data));
+}
+
+template <typename T, bool is_int4 = false>
+void RunQnnTensorImplicitCopyTest(Qnn_DataType_t datatype) {
+  const std::vector<std::uint32_t> kDims = {1, 1, 4};
+  std::vector<T> data = {1, 2, 3, 4};
+  if constexpr (is_int4) {
+    std::vector<int8_t> packed_data;
+    packed_data.reserve(data.size() / 2);
+    for (size_t i = 0; i < data.size() / 2; ++i) {
+      uint8_t low_nibble = data[i * 2] & 0x0F;
+      uint8_t high_nibble = data[i * 2 + 1] & 0x0F;
+      packed_data.emplace_back((high_nibble << 4) | low_nibble);
+    }
+    data = packed_data;
   }
 
-  Qnn_Tensor_t& ref = tensor_wrapper.GetQnnTensor();
-  EXPECT_EQ(ref.version, QNN_TENSOR_VERSION_2);
-  EXPECT_EQ(ref.v2.id, 0);
-  EXPECT_EQ(ref.v2.type, QNN_TENSOR_TYPE_APP_WRITE);
-  EXPECT_EQ(ref.v2.dataFormat, QNN_TENSOR_DATA_FORMAT_FLAT_BUFFER);
-  EXPECT_EQ(ref.v2.dataType, QNN_DATATYPE_UFIXED_POINT_8);
-  EXPECT_EQ(ref.v2.quantizeParams.encodingDefinition, QNN_DEFINITION_UNDEFINED);
-  ASSERT_EQ(ref.v2.rank, dummy_dims.size());
-  for (size_t i = 0; i < ref.v2.rank; i++) {
-    EXPECT_EQ(ref.v2.dimensions[i], dummy_dims[i]);
+  const auto data_size = data.size() * sizeof(T);
+  const void* data_ptr = static_cast<const void*>(data.data());
+
+  TensorWrapper tensor_wrapper{"",       QNN_TENSOR_TYPE_APP_WRITE,
+                               datatype, QuantizeParamsWrapperVariant(),
+                               kDims,    static_cast<uint32_t>(data_size),
+                               data_ptr, false};
+
+  Qnn_Tensor_t cloned_tensor;
+  tensor_wrapper.CloneTo(cloned_tensor);
+  Qnn_Tensor_t& ref_tensor = tensor_wrapper.GetQnnTensor();
+  if constexpr (is_int4) {
+    std::vector<std::int8_t> int8_data;
+    ConvertDataFromInt4ToInt8(data.data(), int8_data, data.size());
+    ValidateTensor(cloned_tensor, QNN_DATATYPE_SFIXED_POINT_8, kDims,
+                   int8_data);
+    ValidateTensor(ref_tensor, QNN_DATATYPE_SFIXED_POINT_8, kDims, int8_data);
+  } else {
+    ValidateTensor(cloned_tensor, datatype, kDims, data);
+    ValidateTensor(ref_tensor, datatype, kDims, data);
   }
-  EXPECT_EQ(ref.v2.memType, QNN_TENSORMEMTYPE_RAW);
-  ASSERT_EQ(ref.v2.clientBuf.dataSize, data_size);
-  const auto* ref_data =
-      static_cast<const std::uint8_t*>(ref.v2.clientBuf.data);
-  for (size_t i = 0; i < data.size(); i++) {
-    EXPECT_EQ(ref_data[i], data[i]);
+}
+
+TEST_P(TensorWrapperDatatypeTest, QnnTensorImplicitCopyTest) {
+  const Qnn_DataType_t datatype = GetParam();
+  switch (datatype) {
```

Comment:
In our current use cases, I think we don't need parameterized tests here because we check the parameter and do different things.

Directly run `RunQnnTensorImplicitCopyTest` with type we want to test.
```suggestion
    RunQnnTensorImplicitCopyTest<std::uint8_t>(QNN_DATATYPE_UFIXED_POINT_8);
    RunQnnTensorImplicitCopyTest<std::int8_t>(QNN_DATATYPE_SFIXED_POINT_8);
    RunQnnTensorImplicitCopyTest<std::uint16_t>(QNN_DATATYPE_UFIXED_POINT_16);
    RunQnnTensorImplicitCopyTest<std::int16_t>(QNN_DATATYPE_SFIXED_POINT_16);
    RunQnnTensorImplicitCopyTest<std::int8_t, true>(QNN_DATATYPE_SFIXED_POINT_4);
```

Reply:
After offline discussion, I have changed to 5 `TEST(TensorWrapperDatatypeTest, <datatype>)` and removed `TEST_P`.
Thank you!